kernel/src/syscall/getdents64.rs

```diff
@@ -146,26 +150,19 @@ impl DirentSerializer for Dirent {
         self.inner.d_reclen as usize
     }
 
-    fn serialize(&self, buf: &mut [u8]) -> Result<()> {
-        // Ensure buffer is large enough for the directory entry
-        if self.len() > buf.len() {
-            return_errno_with_message!(Errno::EINVAL, "buffer is too small");
-        }
+    fn serialize(&self, writer: &mut VmWriter) -> Result<()> {
+        writer.write_val(&self.inner)?;
+
+        let name_bytes = self.name.as_c_str().to_bytes_with_nul();
+        let mut reader = VmReader::from(name_bytes);
+        writer.write_fallible(&mut reader)?;
 
-        let d_ino = self.inner.d_ino;
-        let d_off = self.inner.d_off;
-        let d_reclen = self.inner.d_reclen;
-        let items: [&[u8]; 4] = [
-            d_ino.as_bytes(),
-            d_off.as_bytes(),
-            d_reclen.as_bytes(),
-            self.name.as_c_str().to_bytes_with_nul(),
-        ];
-        let mut offset = 0;
-        for item in items {
-            buf[offset..offset + item.len()].copy_from_slice(item);
-            offset += item.len();
+        let written_len = size_of::<DirentInner>() + name_bytes.len();
+        let padding_len = self.len() - written_len;
+        if padding_len > 0 {
```

Comment:
It seems that `fill_zeros` will check if the length is zero anyway. Do we still need to check it again?
https://github.com/asterinas/asterinas/blob/1aff7d03d19cdc921d4c95d86f8b7958635323e3/ostd/src/mm/io.rs#L790-L793

Reply:
You are absolutely right! No need for this duplicate check.